Add the Miou implementation for caqti

dinosaure opened the pull request:
This is a draft which requires the _upstream_ version of Miou which is not yet released but, as far as I can, it works :+1:. The TLS implementation is not yet done even if a `tls_miou` implementation exists.

paurkedal commented on caqti-miou/lib/system.ml:
```diff
@@ -0,0 +1,237 @@
+
+  let rec terminate orphans =
+    match Miou.care orphans with
+    | None -> ()
+    | Some None -> Miou.yield (); terminate orphans
+    | Some (Some prm) -> (
+        match Miou.await prm with
+        | Ok () -> terminate orphans
+        | Error _exn -> terminate orphans)
+
+  let run fn =
+    let orphans = Miou.orphans () in
+    let hooks = Miou.Sequence.create () in
+    let open Effect.Deep in
+    let retc = Fun.id in
+    let exnc = reraise in
+    let effc : type c. c Effect.t -> ((c, 'r) continuation -> 'r) option =
+      function
+      | Spawn fn ->
+          let _ = Miou.call_cc ~orphans fn in
+          Some (fun k -> continue k ())
+      | _ -> None
+    in
+    let handler = { retc; exnc; effc } in
+    match match_with fn { orphans; hooks } handler with
+    | value -> terminate orphans; release hooks; value
+    | exception exn -> terminate orphans; release hooks; raise exn
+
+  let on_release_cancellable { hooks; _ } fn =
+    Miou.Sequence.(add Left) hooks fn
+
+  let remove_hook hook = Miou.Sequence.remove hook
+
+  let check { orphans; _ } =
+    match Miou.care orphans with
+    | None -> ()
+    | Some None -> ()
+    | Some (Some prm) -> (
+        match Miou.await prm with Ok () -> () | Error _exn -> ())
+end
```
Now, I can see you took the effort of implementing a `Switch` specialized for Miou.  Switches are optional, so if they are not part of the suggested coding style of Miou users, then this is not needed.  In particular, the drivers and other Caqti code does not rely on switches for releasing resources it allocates; if it does, it's a bug.

dinosaure replied:
Unfortunately, the switch is needed. Miou has some contraints such as only a parent can wait its direct children - more concretely, something like `Lwt.async` is strictly forbidden. The idea of the switch is to aggregate all spawned tasks and be sure that they are _awaited_ by the right parent - otherwise, Miou raises a `Miou.Not_a_child` exception.

This switch implementation is here to ensure that the caqti follows these rules in the way it uses Miou.

paurkedal replied:
Aha, I noticed that in the Miou docs.  I'll have a second read with that in mind.  Let me know if you have questions or thoughts on the internal API.  We can make changes there without breaking the public API.

paurkedal commented on caqti-miou/lib/system.ml:
```diff
@@ -0,0 +1,237 @@
+open Caqti_platform
+
+type Caqti_error.msg += Msg_unix of Unix.error * string * string
```
You need to also call `Caqti_error.define_msg` to define a pretty printer.

paurkedal commented on caqti-miou/lib/system.ml:
```diff
@@ -0,0 +1,237 @@
+  let connect_tcp ~sw:_ ~stdenv:_ sockaddr =
+    let socket =
+      match sockaddr with
+      | Unix.ADDR_UNIX _ -> invalid_arg "Caqti_miou.Net.connect_tcp"
```
In the final version, please use the result type here, with a human readable error.

paurkedal reviewed (changes requested):
Thanks for taking the initiative, miou support is welcome.

To get the trivia out of the way, I'll need license headers on all OCaml source files, so that they are recorded by Git along with your contribution.

Running the test case, the binding-based drivers seems to work, but there are issues with pgx.  I suspect a resource leak.

dinosaure commented:
> Thanks for taking the initiative, miou support is welcome.

Thanks, it still is a draft. I try to improve Miou and, at the same time, provide multiple implementations of some libraries. I will set this PR where:
1) Miou will be ready (and released)
2) and when you agree with this PR

> Running the test case, the binding-based drivers seems to work, but there are issues with pgx. I suspect a resource leak.

Is it possible to have a reproducible case to check if I missed something?

dinosaure commented:
> After fixing the URL, the testsuite goes through, but it takes an hour to complete, compared to less that a minute for other variants. Most time is spent in the stream_binary test, which make 65536 individual requests when using the pgx driver. My experience with the other variants is that we really need buffered IO to achieve good performance.

As far as I understand, the `Stream` implementation corresponds to `module Stream = Caqti_platform.Stream.Make (Fiber)`? It will be nice to see how I can optimize the usage of Miou. Actually, the implementation is a bit dumb and did not take advantages of Miou. I updated this PR with your previous comments.

paurkedal commented:
The updates looks good, except I'll need the license headers as well before merging.

I re-tested with sockets, and that's quick, so the slow-down is likely due to the number of packets generated when the output stream is unbuffered.  I don't think the stream implementation should be an issue, since for Miou map and bind are just applications.  (The `stream_binary` test is expected to be slower for the pgx driver than for the postgresql driver, since it does not implement streaming, but a comparison with eio or lwt for this and other tests should be fair.)

The command link I'm using for testing assumes access to a postgresql database.  I usually start it up on the system on my workstation (with a database $USER owned by $USER):
```
dune exec testsuite/main_miou_unix.exe -- -u "pgx:?user=$USER&host=%2fvar%2frun%2fpostgresql" # for socket
dune exec testsuite/main_miou_unix.exe -- -u "pgx://$USER:$PASSWORD@localhost:5432" # for TCP
```
Alternatively add the URL(s) to `testsuite/uris.conf` and run the whole testsuite with`dune runtest`.

But I also have a script to start up a postgresql server as a regular user, which I consider fixing up and committing to the repo.

dinosaure commented:
Hello, after a release cycle for `ocaml-tls`, I've found the time to finish this PR. I rebased the whole PR to have just one commit which contains the various feedbacks you made and I was able to implement the TLS part. This PR should be ready for review and merge after your feedback :+1:.

paurkedal reviewed (changes requested):
Thanks for the update. I'll have a closer look at this, but here are my comments so far.  As mentioned earlier, I think there is a performance hit due to unblocked IO.  I'll make some measurements.

Remember to add license headers to all OCaml source files.  You can keep the copyright or assign it to me.  In the former case I'll try to contact you if I need to change the license or license exception.

dinosaure commented:
I've made a new attempt based on your comments. `eio` seems to have a not-so-obvious dependency path where the core seems to depend on functions such as `gethostbyname` which, it seems to me, inevitably requires a `unix` dependency in my experience.

In this respect, and in view of Miou's design (which is close to that of `lwt`), I've chosen to go down the `lwt` route and propose more or less the same design. I also took the opportunity to reproduce the test you did for `lwt` with regard to the pool.

paurkedal commented:
I agree miou is closer to lwt then eio.  In the latter, system calls goes though a layer of abstraction, meaning a lot of the logic could be implemented in the platform-generic library.  I guess you are referring to the use of `Eio.Net.getaddrinfo_stream`.  Though it's libc in style, I don't see it should require more than a working network stack and configuration of a DNS server.

Please add license headers to all OCaml source files.

Did you manage to test the TLS mode?  The test suite hangs in my case when enabling TLS.  On one run I could see it had gotten through about half of the tests, though.  Let me know if you cannot reproduce it.  Or if you don't have a PostgreSQL database with TLS you can use for testing, in which case I can share a script for setting it up.

Looks good with the new arrangement of the code, though I prefer the unix library to be shipped with the caqti-miou package.  (Originally that was my design decision due to the way lwt was shipped, which I'm sticking to as I want to limit the number of packages.)

paurkedal commented:
This is the script I use: [pgtest_ctl.txt](https://github.com/user-attachments/files/17163051/pgtest_ctl.txt) (without the `.txt` extension). Note the `vardir` in the script, which you may want to adjust.  I store the script in a subdirectory of by working directory, so with the  current setting it will create the database cluster in a subdirectory `_var/lib/postgresql-test`.  After adjusting the script, if needed, run the `initdb` subcommand to configure and initialize a DB cluster, after which the remaining commands can be used to control it.

The failing test in the test suite is `stream_binary`.  It hangs on `pselect6`, and if I stop the DB cluster it goes into a CPU intensive loop involving `clock_nanosleep` and `futex`.

dinosaure commented:
I think this version is the right one. I was able to test with `pgx` and there was a bug in `tls-miou-unix` that we've just fixed. I've set the constraints accordingly. However, I've also changed the way the `Switch` is implemented. It's difficult for me to find a design that's satisfactory in all cases. In this case, I decided to start with the idea that a `Switch` would be a ‘daemon’ (on another domain) to which we would send our tasks (via `async`).

The previous design (_flattening_ the parent link between the place where we wanted to launch the tasks and the main task) could have been interesting, but its behavior tried more to sequence the spawns than to execute them concurrently.

Finally, I buffered the implementation with TLS which made it faster.

paurkedal commented:
It's fine with the rebasing (I have been checking out into separate branches and using `git diff`).

I pushed the miou main for the benchmarks subdirectory.

Did you manage to run with TLS?  I'm still seeing a failure for the `stream_binary` test, though now it terminates with `Parsing_failure "Pgx: result field is too long"`.

The performance looks good now.

paurkedal commented:
About the issue with the switch, I'm wondering if we can do something different.  The only place where `async` is called is when an alarm triggers closing of a connection in `dispose_expiring` in `caqti/lib-platform/pool.ml`.  Automatic evacuation of idle connections can maybe be considered optional, and it's not implemented for the blocking unix variant.  Or maybe it can be done differently.  Ideally it seems to me that ownership of resources involved during this pruning process should belong to the pool.

dinosaure commented:
> Did you manage to run with TLS? I'm still seeing a failure for the stream_binary test, though now it terminates with Parsing_failure "Pgx: result field is too long".

You probably should check your version of `tls-miou-unix`, can you confirm that the version is `1.0.3`?

> About the issue with the switch, I'm wondering if we can do something different. The only place where async is called is when an alarm triggers closing of a connection in dispose_expiring in caqti/lib-platform/pool.ml. Automatic evacuation of idle connections can maybe be considered optional, and it's not implemented for the blocking unix variant. Or maybe it can be done differently. Ideally it seems to me that ownership of resources involved during this pruning process should belong to the pool.

The main problem with Miou (compared with `eio`) is that you can't forget a task that has been launched and only the **direct** parent who launched the task can wait for it. The switch as implemented at this stage simplifies the design by launching a "daemon" which will itself launch the tasks required by `async`.

The old design was a little more complex, where we wanted to ‘flatten’ the `async` to the main task with an effect (so that the direct parent of all these sub-tasks launched by `async` is the main task). The problem was that with the tests I took from `lwt`, we very quickly came to the situation where tasks were launched by sub-tasks, so we also had to attach a handler to these tasks to continue to flatten the `async` towards the main task (even if there was an intermediate task).

It's hard to know at this stage whether one is better than the other, and it may be that the test I've taken from `lwt` ‘overlooks’ the `Switch` implementation (and that, in the end, certain aspects of the test are not particularly expected in a real use of `caqti).

From my point of view, the current desingn has the advantage of being simpler even though it may not necessarily correspond to **every** use case (particularly if you use Miou with no domains). But I think that a little more simplified testing and iterations should be done at this stage and outside of this PR.

paurkedal commented:
> You probably should check your version of `tls-miou-unix`, can you confirm that the version is `1.0.3`?

My mistake, yes, I can confirm the test suite runs through now also with PGX and TLS.

Related to that, is it no possible to avoid the unix depedency for the TLS package?  The reason is that the only driver which use the TLS support is PGX and the main use case for the PGX driver is non-unix platforms, where the PostgreSQL C bindings are not available.

I should mention that the current design of TLS support in Caqti may be revised, if I can find a good way to make a generic implementation based on the pure tls API.  IIRC, the main complication was that I needed to implement buffering in the Caqti code or find a suitable platform-independent implementation.

dinosaure commented:
> Related to that, is it no possible to avoid the unix depedency for the TLS package?

Unfortunately, that's not possible :confused:. In this case, once again, I'm just following the same implementation as `lwt`, which also offers `caqti-tls-lwt`, which requires `unix`.

> I should mention that the current design of TLS support in Caqti may be revised, if I can find a good way to make a generic implementation based on the pure tls API. IIRC, the main complication was that I needed to implement buffering in the Caqti code or find a suitable platform-independent implementation.

Actually, that's what I had to do: buffer `tls-miou-unix`. More generally, you could imagine a `caqti-tls` which would be a _functor_ and which would require a `SOCKET_OPS` and derive this functor according to the scheduler used and the proposed `SOCKET_OPS`.

Derivations of `tls` for schedulers are more or less the same and `tls-mirage` might be what you are looking for: a scheduler/socket independent TLS layer. The abstraction used is `Mirage_flow.S` but you could replace it with `SOCKET_OPS`.

But I think that this subject is outside the scope of this PR :slightly_smiling_face: .

<hr />

It's a little disgression that should take place more in an issue than here, but I've noticed an attempt to abstract the underlying protocol as soon as it's a question of communicating with a database whose purpose is very, very close to [mimic](https://github.com/dinosaure/mimic) (or [provider](https://github.com/mbarbin/provider)).

The two projects above attempt to offer an implementation based on a dynamic resolution mechanism. In this case, mimic can produce an abstract `flow` according to a dynamic context (if TLS is available, according to a pre-registered TCP/IP implementation, according to a given DNS resolution mechanism).

The idea is, as was the case for `ocaml-git` and [paf](https://github.com/dinosaure/paf-le-chien), to use the Mimic interface - [Mirage_flow.S](https://github.com/mirage/mirage-flow/blob/main/src/mirage_flow.mli) - (which only requires `lwt` at this stage but I'm thinking of another level of abstraction) without requiring a TCP/IP, TLS, SSH or HTTP stack.

Next, a ceremony is required for `git-unix`/`paf-unix` or `git-mirage`/`paf-mirage` to register the protocols depending on what is available (`ocaml-tls`, `lwt_ssl`, the `Unix` module, `mirage-tcpip`, `awa-ssh` or `ssh` via `execve`). This ceremony produces a `Mimic.ctx` and, from this, we can obtain a flow in which we can read and write and thus implement the protocol (in this case, for you, the protocol for an SQL database, for me, the Git protocol and the HTTP protocol).

In general, I don't recommend mimic too much because its use is very focused on Mirage (which requires such a high level of abstraction) but, when I was implementing `caqti-miou`, I thought that your codebase could be simplified with mimic - especially as the current caqti design takes ideas from mimic (the heterogeneous table and the first-class modules).

If you're interested, a tutorial is available here: https://dinosaure.github.io/mimic/mimic/index.html

paurkedal commented:
Thanks for the digression and comments, this is precisely the direction I'm hoping for, without reinventing too many wheels.  And yes, I realize we'll need to do miou like lwt given the current design.  So, I think this PR is essentially ready, though I'll take a final look before merging.

Note that I don't indent to release any of TLS packages for the unix platfom until I address the design, and I expect miou users to also prefer the postgresql driver over pgx under unix.

dinosaure commented on .circleci/config.yml:
```diff
@@ -28,6 +28,11 @@ jobs:
+          name: Synchronize with opam-repository
+          command: |
+            opam repo add upstream git+https://github.com/ocaml/opam-repository.git
+            opam update
```
Just a side note that this patch is not needed anymore since `tls-miou-unix.1.0.3` is available from the `opam.ocaml.org` cache. You can delete it :+1:.

paurkedal commented:
Note that if I merge with the current coding style, I will likely do some reformatting if I need to work on the code (esp. the record syntax end the placement of "in").

Apart from the pending issue above (opam file), this should be ready to merge.

dinosaure commented:
Yes, you are free to change the code to suit your style :slightly_smiling_face:  I wasn't sure if it was possible to apply ocamlformat. I've just reverted the change in the CI. From my point of view, everything looks good :+1:.

paurkedal commented:
There is still the packaging issue.

paurkedal commented:
Do you understand why [the pool test failed with a timout in this CI run](https://app.circleci.com/pipelines/github/paurkedal/ocaml-caqti/383/workflows/4a7a944c-e1cc-488e-8310-d53df1eb69d4/jobs/466)?

dinosaure commented:
My last commit should make the test less flaky.

> There is still the packaging issue.

Sorry but what is the issue?

paurkedal commented:
> Sorry but what is the issue?

My mistake, forgot to press the finish-review button.

paurkedal commented on caqti-miou-unix.opam:
```diff
@@ -0,0 +1,27 @@
+  ["dune" "build" "@doc" "-p" name "-j" jobs] {with-doc}
+]
+dev-repo: "git+https://github.com/paurkedal/ocaml-caqti.git"
+synopsis: "Miou support for Caqti"
```
To elaborate, I have bundled the unix packages with the main packages until now, I think originally dune to the fact that lwt does so.  I don't want to change that practise at the moment, since I'm trying to limit the number of packages, and I think it's a good fit for miou, anyway, since it also includes a unix sub-library.

dinosaure commented:
I finally merge `caqti-miou` and `caqti-miou-unix` (and provide `caqti-miou.unix` as `caqti-lwt.unix`). Tell me if it's right, I was not sure about what you expect :+1:.

paurkedal commented:
Yes, that's it, but one last thing (I hope):  I think you should be the author in the opam package, since you wrote most of the code.

paurkedal commented:
Thanks!
